### DISCUSSION TITLE
syntax error with mysql bulk update via INSERT ... SELECT ... ON DUPLICATE KEY UPDATE

### DISCUSSION THREAD

**Discussion by anentropic** [sqlalchemy/sqlalchemy#10514] [Usage Questions]
I am trying to bulk update via an insert on duplicate, but using `from_select` (instead of list of values as seen here https://github.com/sqlalchemy/sqlalchemy/discussions/9328)

I am getting error from mysql like:
```sql
(MySQLdb.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to
your MySQL server version for the right syntax to use near
'AS new ON DUPLICATE KEY UPDATE dealer_dealer_name = new.dealer_dealer_name, cust' at line 3")
[SQL: INSERT INTO mytable (...) SELECT dealer_dealer_names.value AS dealer_dealer_name, ...
FROM mytable
LEFT OUTER JOIN etl_anonymised_company_name AS dealer_dealer_names ON md5(mytable.dealer_dealer_name) = dealer_dealer_names.`key`
...<more similar joins>...
WHERE mytable.last_modified_date >= %s AND mytable.last_modified_date < %s
ORDER BY mytable.last_modified_date
AS new
ON DUPLICATE KEY UPDATE dealer_dealer_name = new.dealer_dealer_name, ...]
```

my sqlalchemy looks like:
```python
    insert_clause = insert(MyTable).from_select(
        [col.name for col in insert_columns],
        make_select_query(insert_columns),
    )
    update_query = (
        insert_clause.on_duplicate_key_update({
            col_name: getattr(insert_clause.inserted, col_name)
            for col_name in update_columns
        })
    )
    with engine.begin() as conn:
        result = conn.execute(update_query)
```

mysql error seems to imply it doesn't like how sqlalchemy has aliased the select query `AS new` (it's not something I have done explicitly in my `select` instance)

if I print `str(update_query)` it looks different, there's no select alias and the field updates look like `ON DUPLICATE KEY UPDATE dealer_dealer_name = VALUES(dealer_dealer_name)`

I realise this isn't a minimal reproducible case at the moment, just wondered if there's something obvious I should be doing differently

**Comment by zzzeek** [comment]
ive never worked with this mysql feature so I'm not familiar with what syntaxes are supported.

to stringify correctly, pass the dialect, `str(stmt.compile(dialect=mysql.dialect()))` see https://docs.sqlalchemy.org/en/20/faq/sqlexpressions.html#stringifying-for-specific-databases

**Reply by CaselIT** [reply]
if I don't remember incorrectly at some point mysql changed syntax and required the use of new.
so @anentropic can you provide the exact version of the db in use, and if it's any dialect. Also if you know the correct query you want it to execute it would be useful to have

**Reply by anentropic** [reply]
Thanks for having a look 😀

the correctly stringified query looks basically the same:

```
(Pdb++) from sqlalchemy.dialects.mysql import dialect
(Pdb++) str(update_query.compile(dialect=dialect()))
'INSERT INTO mytable (...)
SELECT ... FROM mytable
  LEFT OUTER JOIN etl_anonymised_company_name AS dealer_dealer_names
  ON md5(mytable.dealer_dealer_name) = dealer_dealer_names.`key`
  ...
  WHERE ...
  ORDER BY ...
ON DUPLICATE KEY UPDATE dealer_dealer_name = VALUES(dealer_dealer_name), ...'
```

which is a little different from the SQL that MySQL spits out in its error message

I had a play with raw SQL against the db... the form printed by the stringified query above (no select alias, `dealer_dealer_name = VALUES(dealer_dealer_name)` in the update clause) is correct and works fine

so it appears something is changing that into the broken version (`SELECT ... AS new ON DUPLICATE KEY UPDATE dealer_dealer_name = new.dealer_dealer_name`) before it reaches MySQL

**Reply by anentropic** [reply]
I'm on MySQL 8.1.0 macos (homebrew) and SQLAlchemy 2.0.22

**Reply by anentropic** [reply]
and I'm using:
```python
from sqlalchemy import update
from sqlalchemy.dialects.mysql import insert
```

**Reply by CaselIT** [reply]
I've created https://github.com/sqlalchemy/sqlalchemy/issues/10675

**Comment by MarkWine** [comment]
Hey. I've tried making a reproducible example without success. I tried building up tables/queries in similar manner to the code above, including a left outer join and order by in the select clause (the example is obviously a bit nonsensical).

```
from sqlalchemy import String, create_engine, Table, MetaData, Column, Integer

from sqlalchemy.dialects.mysql import insert

metadata = MetaData()

table1 = Table(
    "foo",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("name", String(45)),
)

table2 = Table(
    "bar",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("game", String(45)),
)

engine = create_engine("mysql+mysqlconnector://root:redacted@127.0.0.1/test")

metadata.create_all(engine)

with engine.begin() as conn:
    conn.execute(table1.delete())
    conn.execute(table2.delete())
    insert_table1 = insert(table1).values(
        [
            {"name": "x"},
            {"name": "y"},
            {"name": "w"},
            {"name": "z"},
        ]
    )
    insert_table2 = insert(table2).values(
        [
            {"game": "f"},
            {"game": "g"},
            {"game": "h"},
            {"game": "i"},
        ]
    )
    conn.execute(insert_table1)
    conn.execute(insert_table2)


select_clause = (
    table1.select()
    .where(table1.c.id > 2)
    .outerjoin(table2, table1.c.id == table2.c.id)
    .order_by(table1.c.name)
)
insert_clause = insert(table2).from_select(["id", "game"], select_clause)
update_query = insert_clause.on_duplicate_key_update({"game": "d"})
print(update_query)
with engine.begin() as conn:
    result = conn.execute(update_query)
```
This ran without error and resulted in the expected table based on the following update query output:
```
INSERT INTO bar (id, game) SELECT foo.id, foo.name 
FROM foo LEFT OUTER JOIN bar ON foo.id = bar.id 
WHERE foo.id > %s ORDER BY foo.name ON DUPLICATE KEY UPDATE game = %s
```
The resulting `bar` table looked like:
1	f
2	g
3	d
4	d

I'm guessing there's a little (or lot) of extra going on in that `make_select_query`

Table was based on super simple container with .yml:
```
services:
  db:
    image: mysql
    ports:
      - "3306:3306"
    environment:
      MYSQL_ROOT_PASSWORD: redacted
```
